libs/api/ebpf_api.cpp

```diff
@@ -2242,14 +2256,261 @@ ebpf_program_bind_map(fd_t program_fd, fd_t map_fd)
+            // such, so that it gets freed when the user eventually unsubscribes.
+            auto lock_state = ebpf_lock_lock(&subscription->lock);
+            subscription->async_ioctl_failed = true;
+            ebpf_lock_unlock(&subscription->lock, lock_state);
+            EBPF_RETURN_RESULT(result);
+        } else {
+            // User has canceled subscription. Invoke user specified callback for the final time with NULL record. This
+            // will let the user app clean up its state.
+            subscription->sample_callback(subscription->sample_callback_context, nullptr, 0);
+        }
+    } else {
+        // Async IOCTL operation returned with success status. Read the ring buffer records and indicate it to the
+        // subscriber.
+
+        size_t ring_buffer_size;
+        uint32_t dummy;
+
+        result = _get_map_descriptor_properties(
+            subscription->ring_buffer_map_handle,
+            &dummy,
+            &dummy,
+            &dummy,
+            reinterpret_cast<uint32_t*>(&ring_buffer_size));
+        if (result != EBPF_SUCCESS)
+            EBPF_RETURN_RESULT(result);
+
+        ebpf_operation_ring_buffer_map_async_query_reply_t* reply = &subscription->reply;
+        ebpf_ring_buffer_map_async_query_result_t* async_query_result = &reply->async_query_result;
+        consumer = async_query_result->consumer;
+        producer = async_query_result->producer;
+        for (;;) {
+            auto record = ebpf_ring_buffer_next_record(subscription->buffer, ring_buffer_size, consumer, producer);
+
+            if (record == nullptr)
+                // No more records.
+                break;
+
+            int callback_result = subscription->sample_callback(
+                subscription->sample_callback_context,
+                const_cast<void*>(reinterpret_cast<const void*>(record->data)),
+                record->header.length - EBPF_OFFSET_OF(ebpf_ring_buffer_record_t, data));
+            if (callback_result != 0)
+                break;
+
+            consumer += record->header.length;
+        }
+    }
+
+    // Acquire the subscription lock and within the lock, check if the user has unsubscribed.
+    // If the user has unsubscribed. This is the final callback. Free the subscription context.
+
+    auto lock_state = ebpf_lock_lock(&subscription->lock);
+
+    if (subscription->unsubscribed) {
+        result = EBPF_CANCELED;
+        ebpf_lock_unlock(&subscription->lock, lock_state);
+        _ebpf_ring_buffer_free_subscription(subscription);
+        EBPF_RETURN_RESULT(result);
+    }
+
+    // If still subscribed, post the next async IOCTL call while holding the lock. It is safe to do so as the async call
+    // is not blocking.
+
+    // First, register wait for the new async IOCTL operation completion.
+    result = register_wait_async_ioctl_operation(subscription->async_ioctl_completion);
+    if (result != EBPF_SUCCESS)
+        EBPF_RETURN_RESULT(result);
```

Comment:
Failure to release lock.

Reply:
Since this is a CPP file, it would be nice in the future (not for this PR) to use an RAII model for locking where the unlock is done in a destructor, to avoid bugs like this one.

Reply:
changed to scoped_lock and mutex.